app/scripts/controllers/detect-tokens.js

```diff
@@ -24,71 +23,85 @@ export default class DetectTokensController {
     preferences,
     network,
     keyringMemStore,
+    tokenList,
   } = {}) {
     this.preferences = preferences;
     this.interval = interval;
     this.network = network;
     this.keyringMemStore = keyringMemStore;
+    this.tokenList = tokenList;
+  }
+
+  async _getTokenBalances(tokens) {
+    const ethContract = this.web3.eth
+      .contract(SINGLE_CALL_BALANCES_ABI)
+      .at(SINGLE_CALL_BALANCES_ADDRESS);
+    return new Promise((resolve, reject) => {
+      ethContract.balances([this.selectedAddress], tokens, (error, result) => {
+        if (error) {
+          return reject(error);
+        }
+        return resolve(result);
+      });
+    });
   }
 
   /**
-   * For each token in @metamask/contract-metadata, find check selectedAddress balance.
+   * For each token in the tokenlist provided by the TokenListController, check selectedAddress balance.
    */
   async detectNewTokens() {
     if (!this.isActive) {
       return;
     }
-    if (this._network.store.getState().provider.chainId !== MAINNET_CHAIN_ID) {
+
+    const { tokenList } = this._tokenList.state;
+    if (Object.keys(tokenList).length === 0) {
       return;
     }
 
     const tokensToDetect = [];
     this.web3.setProvider(this._network._provider);
-    for (const contractAddress in contracts) {
+    for (const tokenAddress in tokenList) {
       if (
-        contracts[contractAddress].erc20 &&
-        !this.tokenAddresses.includes(contractAddress.toLowerCase()) &&
-        !this.hiddenTokens.includes(contractAddress.toLowerCase())
+        !this.tokenAddresses.find((address) =>
+          isEqualCaseInsensitive(address, tokenAddress),
+        ) &&
+        !this.hiddenTokens.find((address) =>
+          isEqualCaseInsensitive(address, tokenAddress),
+        )
       ) {
-        tokensToDetect.push(contractAddress);
+        tokensToDetect.push(tokenAddress.toLowerCase());
       }
     }
-
-    let result;
-    try {
-      result = await this._getTokenBalances(tokensToDetect);
-    } catch (error) {
-      warn(
-        `MetaMask - DetectTokensController single call balance fetch failed`,
-        error,
-      );
-      return;
-    }
-
-    tokensToDetect.forEach((tokenAddress, index) => {
-      const balance = result[index];
-      if (balance && !balance.isZero()) {
-        this._preferences.addToken(
-          tokenAddress,
-          contracts[tokenAddress].symbol,
-          contracts[tokenAddress].decimals,
+    const sliceOfTokensToDetect = [];
+    sliceOfTokensToDetect.push(tokensToDetect.slice(0, 1000));
+    sliceOfTokensToDetect.push(
```

Comment:
`push`, cool!  You actually don't need the two `push` directives, you can add multiple items in one go:

```
[].push("hello", "world");
```

We could also initialize the array with those items:

```
const sliceOfTokensToDetect = [
  tokensToDetect.slice(0, 1000),
  tokensToDetect.slice(1000, tokensToDetect.length - 1)
];
```

What is the significance of doing the slicing though?  It's not immediately apparent to me.

Reply:
The size of the data (around 5k) returned from the token service API slows down the token detection, we are splitting the list based on the number of occurrences (assuming the probability of a user having a popular token is higher) so that they can be detected faster followed by the remaining list, rather than waiting on the whole list.